src/Kubernetes.Controller/Certificates/ImmutableCertificateCache.cs

```diff
@@ -0,0 +1,121 @@
+
+    public ImmutableCertificateCache(IEnumerable<TCert> certificates, Func<TCert, IEnumerable<string>> getDomains)
+    {
+        foreach (var certificate in certificates)
+        {
+            foreach (var domain in getDomains(certificate))
+            {
+                if (domain.StartsWith("*."))
+                {
+                    _wildCardDomains.Add(new (domain[1..], certificate));
+                }
+                else
+                {
+                    _certificates[domain] = certificate;
+                }
+            }
+        }
+
+        _wildCardDomains.Sort(DomainNameComparer.Instance);
+    }
+
+
+
+    protected abstract TCert? GetDefaultCertificate();
+
+    public TCert? GetCertificate(string domain)
+    {
+        if (TryGetCertificateExact(domain, out var certificate))
+        {
+            return certificate;
+        }
+        if (TryGetWildcardCertificate(domain, out certificate))
+        {
+            return certificate;
+        }
+
+        return GetDefaultCertificate();
+    }
+
+    protected IReadOnlyList<WildCardDomain> WildcardCertificates => _wildCardDomains;
+
+    protected IReadOnlyDictionary<string, TCert> Certificates => _certificates;
+
+    protected record struct WildCardDomain(string Domain, TCert? Certificate);
+
+    private bool TryGetCertificateExact(string domain, [NotNullWhen(true)] out TCert? certificate) =>
+        _certificates.TryGetValue(domain, out certificate);
+
+    private bool TryGetWildcardCertificate(string domain, [NotNullWhen(true)] out TCert? certificate)
+    {
+        if (_wildCardDomains.BinarySearch(new WildCardDomain(domain, null!), DomainNameComparer.Instance) is { } index and > -1)
+        {
+            certificate = _wildCardDomains[index].Certificate!;
+            return true;
+        }
+
+        certificate = null;
+        return false;
+    }
+
+    /// <summary>
+    /// Sorts domain names right to left.
+    /// This allows us to use a Binary Search to achieve a suffix
+    /// search.
+    /// </summary>
+    private class DomainNameComparer : IComparer<WildCardDomain>
+    {
+        public static readonly DomainNameComparer Instance = new();
+
+        public int Compare(WildCardDomain x, WildCardDomain y)
+        {
+            var ret = Compare(x.Domain.AsSpan(), y.Domain.AsSpan());
+            if (ret != 0)
+            {
+                return ret;
+            }
+
+            return (x.Certificate, y.Certificate) switch
+            {
+                (null, not null) when x.Domain.Length > y.Domain.Length => 0,
+                (not null, null) when x.Domain.Length < y.Domain.Length => 0,
+                _ => x.Domain.Length - y.Domain.Length
+            };
+        }
+
+        private static int Compare(ReadOnlySpan<char> x, ReadOnlySpan<char> y)
+        {
+
+            var length = Math.Min(x.Length, y.Length);
+
+            for (var i = 1; i <= length; i++)
```

Comment:
So the JIT can't elide the bound-check for indexing into the spans.
I think it's better to slice the spans to the common (= min) length and use one span in the loop, in order to elide at least one bound check (the JIT should handle the downward loop (see other comment) in a recent version).

Roughly:
```suggestion
            if (x.Length < y.Length)
            {
                y = y.Slice(0, x.Length);
            }
            else if (y.Length < x.Length)
            {
                x = x.Slice(0, y.Length);
            }

            for (var i = x.Length; i >= 1; --i)
```